File: proyecto/cppcalc/ast.cpp
#include <sstream>
#include <string>
#include "ast.h"
#include <iostream>
#include "calculator.h"

using namespace std;

// for debug information uncomment
// #define debug

AST::AST() {}

AST::~AST() {}

BinaryNode::BinaryNode(AST* left, AST* right):
   AST(),
   leftTree(left),
   rightTree(right)
{}

BinaryNode::~BinaryNode() {
#ifdef debug
   cout << "In BinaryNode destructor" << endl;
#endif

   try {
      delete leftTree;
   } catch (...) {}

   try {
      delete rightTree;
   } catch(...) {}
}
   
AST* BinaryNode::getLeftSubTree() const {
   return leftTree;
}

AST* BinaryNode::getRightSubTree() const {
   return rightTree;
}

UnaryNode::UnaryNode(AST* sub):
   AST(),
   subTree(sub)
{}

UnaryNode::~UnaryNode() {
#ifdef debug
   cout << "In UnaryNode destructor" << endl;
#endif
   
   try {
      delete subTree;
   } catch (...) {}

}
   
AST* UnaryNode::getSubTree() const {
   return subTree;
}

AddNode::AddNode(AST* left, AST* right):
   BinaryNode(left,right)
{}

int AddNode::evaluate() {
   return getLeftSubTree()->evaluate() + getRightSubTree()->evaluate();
}

SubNode::SubNode(AST* left, AST* right):
   BinaryNode(left,right)
{}

int SubNode::evaluate() {
   return getLeftSubTree()->evaluate() - getRightSubTree()->evaluate();
}

TimesNode::TimesNode(AST* left, AST* right):
   BinaryNode(left,right)
{}

int TimesNode::evaluate() {
   return getLeftSubTree()->evaluate() * getRightSubTree()->evaluate();
}

DivideNode::DivideNode(AST* left, AST* right):
   BinaryNode(left,right)
{}

int DivideNode::evaluate() {
   return getLeftSubTree()->evaluate() / getRightSubTree()->evaluate();
}

//Modulo

ModuloNode::ModuloNode(AST* left, AST* right):
   BinaryNode(left,right)
{}

int ModuloNode::evaluate() {
   return getLeftSubTree()->evaluate() % getRightSubTree()->evaluate();
}

StoreNode::StoreNode(AST* sub):
  UnaryNode(sub)
{}

int StoreNode::evaluate(){
  calc->store(getSubTree()->evaluate());
  return calc->recall();
}

NumNode::NumNode(int n) :
   AST(),
   val(n)
{}

int NumNode::evaluate() {
   return val;
}

//Creacion y evaluacion del nodo Minus

MinusNode::MinusNode(AST* sub):
  UnaryNode(sub)
{}

int MinusNode::evaluate(){
  calc->minus(getSubTree()->evaluate());
  return calc->recall();
}

//Creacion y evaluacion del nodo Plus

PlusNode::PlusNode(AST* sub):
  UnaryNode(sub)
{}

int PlusNode::evaluate(){
  calc->plus(getSubTree()->evaluate());
  return calc->recall();
}

//El AssignNode covierte el flujo de int a str para que se pueda concatenar

AssignNode::AssignNode(string id, AST* sub):
  UnaryNode(sub),
  id(id)
{}

int AssignNode::evaluate(){
  int val = getSubTree()->evaluate();
  string assignation;

//flujo
  stringstream convert;

  convert<<val;
  assignation = convert.str();

  calc->insert(id,val);
  calc->pushAssign(id + " <- " + assignation);

  return val;
}


IdentifierNode::IdentifierNode(string id):
  AST(),
  id(id)
{}

int IdentifierNode::evaluate(){
  return calc->obtMemory2(id);
}

RecallNode::RecallNode():
  AST()
{}

int RecallNode::evaluate(){
  return calc->recall();
}

ClearNode::ClearNode():
  AST()
{}

int ClearNode::evaluate(){
  calc -> clear();
  int a=0;
  return a;
}

File: proyecto/rubycalc/lib/ast.rb
require 'set'

#El @ es una variable instancia, un atributo de la clase
#Elreturn ni los () son necesarios en ruby

class BinaryNode
  attr_reader :left, :right
  
  def initialize(left,right)
    @left = left
    @right = right
  end
end

class UnaryNode
  attr_reader :subTree
  
  def initialize(subTree)
    @subTree = subTree
  end
end

class ListExprNode < BinaryNode
  def initialize(left,right)
    @left = left
    @right = right
  end

  def evaluate
    if(@left != nil)
      @left.evaluate
    end
    $calc.add(@right.evaluate)
  end

end

class AssignNode < UnaryNode
  def initialize(id,subTree)
    super(subTree)
    @id=id
  end

  def evaluate
    val = @subTree.evaluate
    $calc.insert(@id,val)

    #Asigna
    return val.to_i
  end
end

class IdentifierNode
  def initialize(id)
    @id=id
  end

  def evaluate
    return $calc.search(@id).to_i
  end
end

#to_i convierte el token a integer

class AddNode < BinaryNode
  def initialize(left, right)
    super(left,right)
  end
  
  def evaluate()
    return @left.evaluate.to_i + @right.evaluate.to_i
  end
end

class SubNode < BinaryNode
  def initialize(left, right)
    super(left,right)
  end
  
  def evaluate() 
    return @left.evaluate.to_i - @right.evaluate.to_i
  end
end

class NumNode
  def initialize(num)
    @num = num
  end

  def evaluate()
    @num
  end
end

class StoreNode < UnaryNode
  def initialize(subTree)
    super(subTree)
  end

  def evaluate()
    $calc.memory = subTree.evaluate.to_i
  end
end

class PlusNode < UnaryNode
  def initialize(subTree)
    super(subTree)
  end

  def evaluate
    $calc.memory+=subTree.evaluate.to_i
  end
end

class MinusNode < UnaryNode
  def initialize(subTree)
    super(subTree)
  end

  def evaluate
    $calc.memory-=subTree.evaluate.to_i
  end
end

class ClearNode
  def evaluate
    $calc.memory2.clear
    $calc.memory=0
  end
end

class RecallNode
  def evaluate
    $calc.memory.to_i
  end
end

class TimesNode < BinaryNode
  def initialize(left, right)
    super(left,right)
  end
  
  def evaluate 
    return @left.evaluate.to_i * @right.evaluate.to_i
  end
end

class DivideNode < BinaryNode
  def initialize(left, right)
    super(left,right)
  end
  
  def evaluate()
    return @left.evaluate.to_i / @right.evaluate.to_i
  end
end

class ModuloNode < BinaryNode
  def initialize(left, right)
    super(left,right)
  end
  
  def evaluate 
    return @left.evaluate.to_i % @right.evaluate.to_i
  end
end
